test/e2e_node/node_pid_limit_test.go

```diff
@@ -0,0 +1,179 @@
+	Context(fmt.Sprintf("with limit of %d available PIDs", desiredPIDSSmall), func() {
+		tempSetCurrentKubeletConfig(f, setupNodePidLimitTest(desiredPIDSSmall))
+		runNodePidLimitTest(f, true, true, pidConsumingPods("single-big-pod-fail", 1, 150, 60))
+		runNodePidLimitTest(f, true, true, pidConsumingPods("multi-big-pods-fail", 3, 150, 60))
+		runNodePidLimitTest(f, false, false, pidConsumingPods("single-pod-succeed", 1, 50, 60))
+		runNodePidLimitTest(f, true, false, pidConsumingPods("multi-small-pods-fail", 2, 50, 60))
+		runNodePidLimitTest(f, false, false, pidConsumingPods("multi-verysmall-pods-succeed", 2, 30, 60))
+	})
+})
+
+func runNodePidLimitTest(f *framework.Framework, shouldFailSome, shouldFailAll bool, pods []*v1.Pod) {
+	Context("", func() {
+		var npods []*v1.Pod
+		BeforeEach(func() {
+			framework.Logf("Creating pods %#+v", pods)
+			npods = f.PodClient().CreateBatch(pods)
+		})
+		It("checks for appropriate pod status", func() {
+			By("Waiting for pods to terminate")
+			podSuccessCount := 0
+			podFailCount := 0
+			var podErrors []string
+			for _, pod := range npods {
+				framework.Logf("Waiting for pod %s, %v", pod.ObjectMeta.Name, podCompletionTimeout)
+				_ = framework.WaitTimeoutForPodNoLongerRunningInNamespace(f.ClientSet, pod.ObjectMeta.Name, f.Namespace.Name, podCompletionTimeout)
+				framework.Logf("    Waited for pod %s/%s", f.Namespace.Name, pod.ObjectMeta.Name)
+			}
+			updatedPodsList, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).List(metav1.ListOptions{})
+			if err != nil {
+				Fail(fmt.Sprintf("Failed to retrieve pod status: %s", err))
+			}
+			framework.Logf("  All pods done:")
+			for _, pod := range updatedPodsList.Items {
+				if pod.Status.Phase == v1.PodSucceeded {
+					framework.Logf("    Pod %s succeeded", pod.ObjectMeta.Name)
+					podSuccessCount++
+				} else if pod.Status.Phase == v1.PodFailed {
+					framework.Logf("    Pod %s failed", pod.ObjectMeta.Name)
+					podFailCount++
+				} else {
+					framework.Logf("    Pod %s did something weird (%v)", pod.ObjectMeta.Name, pod.Status.Phase)
+					podErrors = append(podErrors, fmt.Sprintf("Unexpected status from %s: %v", pod.ObjectMeta.Name, pod.Status.Phase))
+				}
+			}
+			if len(podErrors) > 0 {
+				Fail(fmt.Sprintf("Errors observed: %s", strings.Join(podErrors, ", ")))
+			} else if shouldFailAll && podSuccessCount > 0 {
+				Fail(fmt.Sprintf("Expecting all pods to fail, but %d succeeded", podSuccessCount))
+			} else if shouldFailSome && podFailCount == 0 {
+				Fail(fmt.Sprintf("Expecting some pods to fail, but none did"))
+			} else if !shouldFailSome && !shouldFailAll && podFailCount > 0 {
+				Fail(fmt.Sprintf("Expecting all pods to succeed, but %d failed", podFailCount))
+			}
+		}, 600.0)
+		AfterEach(func() {
+			for _, pod := range npods {
+				f.PodClient().DeleteSync(pod.Name, &metav1.DeleteOptions{}, 2*time.Minute)
+			}
+		})
+	})
+}
+
+const (
+	pidPodCommand = `
```

Comment:
to make sure I understand, this command will exit if it can't fork, right?  And we wait for all the pods to fail as evidence that they hit the allocatable limit?

Reply:
> to make sure I understand, this command will exit if it can't fork, right?
Yes.  And it will also exit if it reaches the configured limit (in that case, it will exit status 0 for success).  But there's a timer in there, to ensure that the pod's processes will stick around long enough to let any other pods being run as part of the test do their job.

>And we wait for all the pods to fail as evidence that they hit the allocatable limit?
Unless each pod (in addition to anything else that happens to be running in that cgroup) is individually over the limit, there's no guarantee that they all will fail; it's a matter of timing.

For example, if the threshold is 100 pids and it starts 2 pods each configured to fork 60 times, one or both pods will fail, but it's not guaranteed that both will.  For example, if the first pod completes all of its forks (and then sticks around for a minute), it will succeed, but the second pod will fail earlier.

The test is written such that there can be three outcomes: all succeed (if the total is less than the limit), all fail (if each pod is individually above the limit), or some fail (if each pod is individually below the limit, but the sum is above the limit).

Reply:
> The test is written such that there can be three outcomes: all succeed (if the total is less than the limit), ...

Does that mean that if the feature is broken (e.g. doesn't set the cgroup value) the test will still pass?  Or are we making sure a certain number of pods fail?

Reply:
If the number of PIDs individually is below the limit, but the total is above the limit, we can't be sure just how many pods will fail.

For example: suppose we have a limit of 100 process IDs (excluding other pods, for the sake of discussion) and each pod uses 20, and we create 6  pods.  It's possible for anywhere from 1 to 6 pods to fail, depending upon the exact timing of how the pods are run.  For example, if each pod happens to be run in order, only the last pod will fail.  If they are run concurrently, and round-robin their forks, then all will fail.  We cannot guarantee how many will fail, but we can guarantee that some will.  I suppose we could refine it to say "expect this minimum number to fail", where a positive number is the lower bound on the number of failures, 0 means none are expected to fail, and -1 indicates that all are expected to fail, or some such.

Reply:
Instead of running more pods, could we instead run a single pod that consumes a certain number of pids?  Then we just expect the pod to fail or succeed based on if it is set to use more/less that the number of allocatable pids.  This sounds like it might be hard to debug if, for example, the test is flaky for some reason.  I wouldn't know if it was because of ordering or a kubelet bug...

Reply:
Certainly we can do it that way.  The reason I prefer to do it this way is to distinguish between a PID limit applied per-pod and one applied to all pods.  If, for example, there were a bug in the cgroup code where it inadvertently set the limit on the pod's cgroup rather than top-level one, the test would pass even though the functionality was broken.  The way I'm doing it we would see the test fail.

The test creates the same number of process IDs either way, and tries to sleep for long enough (1 minute) to ensure that both pods have time to run before exiting.

Reply:
I see your point.  Historically, non-deterministic tests have been a much larger problem than those kinds of bugs though.  Plus, we have the cgroup inspection test as well to make sure we change the right cgroup values.  I would be fine with anything where I know what is supposed to happen to each pod.

Reply:
Fair enough.  There's no simple way (it would require coordination in time between the two pods) to test multiple pods individually below the limit but collectively above it.  I've made this change.